src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -759,17 +753,32 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                 throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(SslStream.ReadAsync), "read"));
             }
 
+            ThrowIfExceptionalOrNotAuthenticated();
+
             Debug.Assert(_internalBuffer is null || _internalBufferCount > 0 || _decryptedBytesCount > 0, "_internalBuffer allocated when no data is buffered.");
 
             try
             {
-                while (true)
+
+                if (_decryptedBytesCount != 0)
                 {
-                    if (_decryptedBytesCount != 0)
-                    {
-                        return CopyDecryptedData(buffer);
-                    }
+                    int length = CopyDecryptedData(buffer);
+                    ReturnReadBufferIfEmpty();
+                    return length;
+                }
 
+                if (receivedEOF)
+                {
+                    // We received EOF during previous read but had buffered data to return.
+                    return 0;
+                }
+
+                Debug.Assert(_decryptedBytesOffset == 0);
+                Debug.Assert(_decryptedBytesCount == 0);
+
+                int processedLength = 0;
+                while (true)
+                {
```

Comment:
I don't think we want the zero-byte read inside the loop here, do we? We just want to do it once.

Reply:
Yes, in concept. The  #49743 took it out but then I had some troubles with some of the tests.  Since I'm trying to break the change to smaller chunks I decided to keep it where it is right now. I can take that as follow-up.

Reply:
This got fixed right? Just want to confirm...

Reply:
yes.